src/func.rs

```diff
@@ -139,24 +129,43 @@ impl FuncInstance {
 	///
 	/// [`signature`]: #method.signature
 	/// [`Trap`]: #enum.Trap.html
-	pub fn invoke<E: Externals>(
+	pub fn invoke_configurable<E: Externals>(
 		func: &FuncRef,
 		args: &[RuntimeValue],
 		externals: &mut E,
+		interpreter: &mut Interpreter,
 	) -> Result<Option<RuntimeValue>, Trap> {
 		check_function_args(func.signature(), &args)?;
 		match *func.as_internal() {
 			FuncInstanceInternal::Internal { .. } => {
-				let mut interpreter = Interpreter::new(func, args)?;
-				interpreter.start_execution(externals)
+				interpreter.start_execution(externals, func, args)
 			}
 			FuncInstanceInternal::Host {
-				ref host_func_index,
-				..
+				ref host_func_index, ..
 			} => externals.invoke_index(*host_func_index, args.into()),
 		}
 	}
 
+	/// Invoke this function.
+	///
+	/// # Errors
+	///
+	/// Returns `Err` if `args` types is not match function [`signature`] or
+	/// if [`Trap`] at execution time occured.
+	///
+	/// [`signature`]: #method.signature
+	/// [`Trap`]: #enum.Trap.html
+	pub fn invoke<E: Externals>(
```

Comment:
Maybe remove this entirely and have `invoke` be a method on `Interpreter`? Combined with my comment below that would mean invoking a function with a new interpreter would be `Interpreter::new().invoke(func, args, ext)` - much more explicit about the fact that you're creating a new `Interpreter`.

Reply:
Yeah, that's nice and ergonomic. I went ahead and moved `invoke` into `Interpreter`.

Perhaps `invoke_resumable` should be moved over as well?

Reply:
Yeah, I think that would be sensible. Could you do it as part of this PR, please? It doesn't make sense to leave it half-done.

Reply:
Starting on that now.